File: src/Auth/AuthException.php
<?php

declare(strict_types=1);

namespace MiniS3\Auth;

use RuntimeException;

final class AuthException extends RuntimeException
{
    public function __construct(
        private readonly string $s3Code,
        string $message,
        private readonly int $httpStatus = 401
    ) {
        parent::__construct($message);
    }

    public function getS3Code(): string
    {
        return $this->s3Code;
    }

    public function getHttpStatus(): int
    {
        return $this->httpStatus;
    }
}

File: go/internal/auth/auth_exception.go
package auth

import "fmt"

// Exception is the Go equivalent of MiniS3\Auth\AuthException. It carries
// an S3 error code and an HTTP status alongside the error message.
type Exception struct {
	S3Code     string
	Message    string
	HTTPStatus int
}

func (e *Exception) Error() string {
	return fmt.Sprintf("%s: %s", e.S3Code, e.Message)
}

// NewException creates an auth Exception with the default HTTP status 401.
func NewException(s3Code, message string) *Exception {
	return &Exception{S3Code: s3Code, Message: message, HTTPStatus: 401}
}

// NewExceptionWithStatus creates an auth Exception with a custom HTTP status.
func NewExceptionWithStatus(s3Code, message string, httpStatus int) *Exception {
	return &Exception{S3Code: s3Code, Message: message, HTTPStatus: httpStatus}
}
